torch/_dynamo/__init__.py

```diff
@@ -112,3 +112,49 @@ def fn(a):
 def graph_break():
     """Force a graph break"""
     pass
+
+
+def forbid_in_graph(fn):
+    """
+    Customize which functions TorchDynamo will assert are not present while tracing.
+
+    If you want a graph break on this function instead, use disallow_in_graph.
+    TODO(voz): We now have allow_in_graph, disallow_in_graph, forbid_in_graph - some more robust
+    documentation would not be amiss.
+    """
+    if isinstance(fn, (list, tuple)):
+        return [forbid_in_graph(x) for x in fn]
+    assert callable(fn), "forbid_in_graph applies only to callables"
+    fn._dynamo_forbidden = True
+    return fn
+
+
+@forbid_in_graph
+def mark_dynamic(t, index):
+    """
+    Mark a tensor as having a dynamic dim.
+
+    The behavior of having a dynamic dimension on a tensor is governed by a few factors:
+    1) torch._dynamo.config dynamic_shapes True or False - we do not support False atm, but will in the future.
+    dynamic_shapes must be True for mark_dynamic to work.
```

Comment:
The phrasing of this suggests we should raise an error if you mark dynamic and dynamic shapes is not True. But this is not the case; I can mark without the config and then pass it to dynamic=True region no-problemo. So seems like it needs some rewording.

There's another question, which is if a NON dynamic torch.compile region gets a tensor marked as having dynamic dim, what should we do? Erroring here seems good

Reply:
> The phrasing of this suggests we should raise an error if you mark dynamic and dynamic shapes is not True. But this is not the case; 
> There's another question, which is if a NON dynamic torch.compile region gets a tensor marked as having dynamic dim, what should we do? Erroring here seems good

Is that not the test `test_no_dynamic_shapes_mark_dynamic_illegal` ? Perhaps I am misunderstanding what you mean? Can you elaborate?

Reply:
I'm saying that the docs here are confusing. You are indeed correct that my second question is what the code currently does. I did not understand that from the docs here.

Reply:
Let me try to rephrase this.